File: Programmers/연습문제/Level 1/체육복/Solution.js
function solution(n, lost, reserve) {
    let gymSuits = setGymSuits(n, lost, reserve);
    borrowReserve(gymSuits);
    const answer = setAnswer(gymSuits);
    return answer;
}

function setGymSuits(n, lost, reserve) {
    let gymSuits = new Array(n).fill(1);
    for (const i of lost) {
        gymSuits[i-1]--;
    }
    for (const i of reserve) {
        gymSuits[i-1]++;
    }
    return gymSuits;
}

function borrowReserve(gymSuits) {
    for (let i=0; i<gymSuits.length; i++) {
        if (gymSuits[i] == 2) {
            if (i != 0 && gymSuits[i-1] == 0) {
                gymSuits[i]--;
                gymSuits[i-1]++;
            } else if (i != gymSuits.length - 1 && gymSuits[i+1] == 0) {
                gymSuits[i]--;
                gymSuits[i+1]++;
            }
        }
    }
}

function setAnswer(gymSuits) {
    return gymSuits.filter(a => a>=1).length;
}


// TC
const n = 4;
const lost = [3, 1]; 
const reserve = [2, 4];
console.log(solution(n, lost, reserve)); // 5

File: Programmers/연습문제/Level 1/체육복/Solution.swift
import Foundation

func solution(_ n:Int, _ lost:[Int], _ reserve:[Int]) -> Int {
    let gymsuits = setGymsuits(n, lost, reserve)
    let answer = borrowReserve(n, gymsuits)
    return answer
}

func setGymsuits(_ n: Int, _ lost: [Int], _ reserve: [Int]) -> [Int] {
    var gymsuits = Array(repeating: 1, count: n)
    for i in lost {
        gymsuits[i-1] -= 1
    }
    for j in reserve {
        gymsuits[j-1] += 1
    }
    return gymsuits
}

func borrowReserve(_ n: Int, _ gymsuits: [Int]) -> Int {
    var gymsuits = gymsuits
    for i in 0..<n {
        if gymsuits[i] >= 2 {
            if i != 0 && gymsuits[i-1] == 0 {
                gymsuits[i] -= 1
                gymsuits[i-1] += 1
            } else if i != n-1 && gymsuits[i+1] == 0 {
                gymsuits[i] -= 1
                gymsuits[i+1] += 1
            }
        }
    }
    return gymsuits.filter{ $0 >= 1 }.count
}

    
// TC
let n = 5
let lost = [2, 4]
let reserve = [1, 3, 5]
print(solution(n, lost, reserve)) // 5
